contracts/oracles/yieldspace/PoolOracle.sol

```diff
@@ -0,0 +1,146 @@
+        return epochPeriod % granularity;
+    }
+
+    /// @dev returns the observation from the oldest epoch (at the beginning of the window) relative to the current time
+    /// @param pool Address of pool for which the observation is required
+    /// @return The oldest observation available for `pool`
+    function getOldestObservationInWindow(address pool) public view returns (Observation memory) {
+        if (poolObservations[pool].length == 0) {
+            revert NoObservationsForPool(pool);
+        }
+
+        unchecked {
+            uint256 observationIndex = observationIndexOf(block.timestamp);
+            uint256 length = poolObservations[pool].length;
+            uint256 oldestObservationIndex;
+            // can't possible overflow
+            for (uint256 i; i < length; ++i) {
+                // no overflow issue. if observationIndex + 1 overflows, result is still zero.
+                oldestObservationIndex = (++observationIndex) % granularity;
+                // If no data exists for this index, check the next one,
+                // this should only happen during the first timeWindow for a given pool
+                // If the elapsedTime is bigger than the timeWindow,
+                // we also check the next one in case only this particular observation is stale
+                if (block.timestamp - poolObservations[pool][oldestObservationIndex].timestamp < windowSize) {
+                    return poolObservations[pool][oldestObservationIndex];
+                }
+            }
+
+            revert MissingHistoricalObservation(pool);
+        }
+    }
+
+    // @inheritdoc IPoolOracle
+    function update(address pool) public override {
+        // populate the array with empty observations (oldest call only)
+        for (uint256 i = poolObservations[pool].length; i < granularity; i++) {
+            poolObservations[pool].push();
+        }
+
+        // get the observation for the current period
+        uint256 index = observationIndexOf(block.timestamp);
+        Observation storage observation = poolObservations[pool][index];
+
+        // we only want to commit updates once per period (i.e. windowSize / granularity)
+        uint256 timeElapsed = block.timestamp - observation.timestamp;
+        if (timeElapsed > periodSize) {
+            observation.timestamp = block.timestamp;
+            observation.ratioCumulative = _getCurrentCumulativeRatio(pool);
+            emit ObservationRecorded(pool, index, observation);
+        }
+    }
+
+    /// @inheritdoc IPoolOracle
+    function peek(address pool) public view override returns (uint256 twar) {
+        Observation memory oldestObservation = getOldestObservationInWindow(pool);
+
+        uint256 timeElapsed = block.timestamp - oldestObservation.timestamp;
+        if (timeElapsed > windowSize) {
+            revert MissingHistoricalObservation(pool);
+        }
+
+        if (timeElapsed < minTimeElapsed) {
```

Comment:
so this fn will always revert during the first window and also I guess if there was an entire window that passed without any update being called

Reply:
so i guess if the oracle is in this state the solution is to call `update` then wait the minTimeElapsed before this oracle is useable...right?

Reply:
Yes, this is an extra safeguard I added after I did the "walk-thorough" fallback mechanism on the circular buffer, the original code would fail on a missed observation and you'd have to wait at least 1 slot to get it to work again (assuming the next slot was correct). That'd also mean that the first time you use it for a given pool you'd have to wait 24hs for the thing to be usable. My impl only needs 5 min after it's initialised, which in our particular case would be 5min after we do `setSource` on the price oracle (the one that consumes this one)

Reply:
cool